### ISSUE TITLE
IP block on too many errors

### ISSUE THREAD

**Issue by lifeforms** [SpiderLabs/owasp-modsecurity-crs#552]
Watch for the following events:
- CRS blocked transaction
- 403 response from backend
- 404 response from backend
- 5xx response from backend?

Increment a counter for the client IP address, and block if a threshold is reached.

The WordPress plugin "iThemes Security" has a terrific feature where it bans a client IP if they cause too many 404 errors. This is great in combination with ModSec, as all further requests from a scanner will get a 403 response from the backend, and thus are logged in the ModSecurity audit log. A lot of bots get tripped up by this. This provides a lot of insight in attack techniques that the CRS is not currently detecting, as you get logs for all the vuln scanning requests (after the limit is tripped).

We could make counter rules using a similar approach as the existing optional DOS rules in the CRS.

Users could set thresholds, like 5 errors in a 60 sec period.

We could make tweakable _which_ types of events we should watch. I've seen rare FP with 404 detection when a site has some broken links. So, maybe the default should be to only watch CRS blockings + 403 responses from the backend, and disable 404 watching.

**Comment by dune73** [comment]
Generally CRS looks at every request individually.

CRS3 breaks this policy with the various triggers of the `IP.block` variable (which I do not really like). Then of course the newly mainstreamed DoS rules.

This issue pushes further in this direction.

It does make sense. But I miss the big strategy behind it. And I do not like your use-case.
If you just want the audit.log entries, then write a rule that triggers IP.do_audit_log when a threshold is reached / condition is met and then enable the auditlog based on this variable.

If we can come up with a strategy for CRS looking at complete sessions (vs. individual requests), then I am in. But I am reluctant to add session-based blocking stuff based on some limited heuristic.

And finally a detail: I really enjoy path whitelists returning 404 if a certain uri is not allowed. So if a client trips into 404 too often, he should continue to receive 404s. Not 403s.

**Comment by rm3nchaca** [comment]
I was facing this problem  a few years ago too,  and I solved it with a lua script under nginx.
https://gist.github.com/rm3nchaca/92cad42270f7ae809379

Hope that helps you in the meantime!

**Comment by csanders-git** [comment]
@lifeforms I think the idea is novel perhaps we can add this into the DOS rules pretty easily. @dune73 brings up some good points --- mostly about 404 and 403 being complicated in this case. I think no one would be upset if we added to the DOS score in the event a 500 level error was isused -- no?

**Comment by lifeforms** [comment]
@rm3nchaca Thanks! For the CRS we can't depend on Lua unfortunately, but I might play with your script :)

@dune73 Well I think 'only audit log' is actually a possibly very interesting addition, however I also like to just be able to just block a client after they caused X errors. I've been experimenting with the httpBL, and most bots that I find are not in their blocklist unfortunately. So the CRS doen't have a real good bot story, but iThemes finds a lot of them. I would not advocate enabling this for everyone by the way! It can cause hard-to-diagnose blockings, and this should be limited to people explicitly enabling the feature. Just something very similar to how the DoS rules are optionally configured right now, maybe like:

```
#SecAction \
# "id:'900710',\
#  phase:1,\
#  nolog,\
#  pass,\
#  t:none,\
#  setvar:'tx.error_burst_time_slice=120',\
#  setvar:'tx.error_counter_threshold=10',\
#  setvar:'tx.error_watch_anomaly_score=1',\
#  setvar:'tx.error_watch_403=1',\
#  setvar:'tx.error_watch_404=0',\
#  setvar:'tx.error_watch_5xx=0'"
```

@csanders-git For performance reasons it's probably a wise idea to use the existing DOS blocking logic (ip.block), but I feel that it would be more logical to use a separate counter. Fortunately for almost all normal clients the rate of errors should be quite low, so I'm not very worried about adding another counter. But I haven't used the v3 DOS rules yet.

**Comment by dune73** [comment]
@lifeforms This is a possible solution for an implementation, but I do not see a big strategy. This would bring us to 3 separate session-blocking-mechnims:
* DoS blocking
* UA rules and friends -> ip.block
* new response status blacklist

All somehow similar and still separate. I really think we need a strategy before we add new isolated and hard to debug session blockings.

I am also worried because of clients running into this trap by accident. Yesterday, I saw a broken page in a CMS which resulted in 12 404s in a single page load. Page layout completely broken. `Reload` seems to be the natural resort and in no time, you add 30-40-50 404s on a legitimate client.

To sum it up: I am not completely against this and I actually see the benefit, which is potentially big, but I do not like to add another separate mechanism without an overall strategy.

**Comment by lifeforms** [comment]
@dune73 Yeah, misconfigured backends will really wreak havoc with this feature, so the blocking time should probably be sufficiently small. Not sure what would be involved with a bigger strategy. I see a bigger role for the CRS in dealing with botnets in the far future - been thinking of using rbl on block so people can report blocked IPs to us automatically and create our crowdsourced BL :) Or maybe feed data to httpBL (they don't support that now). But that would be far away. For now I'll just drop the "sticky blocking" idea but I can sleep on it :)

**Comment by dune73** [comment]
Don't drop this too fast. But sleeping on it sounds like a good plan.

From my point of view, an overall strategy has the following characteristics:
- a single mechanism to block client IPs / sessions
- various sensor / detection mechanisms can trigger the said blocking mechanism
- The relationship with the scoring mechanism is clear / sorted out. I think scoring should lead to IP / session blocking eventually.
- This is all designed in transparent way so that casual admins find their way around.
- I am dreaming of an elegant design. Elegant like the new crs-setup.conf and the separate default values.

**Comment by lifeforms** [comment]
@dune73 Oh I meant drop as in "I'll leave this here". I like your points. I'll return to it in due time!

**Comment by dune73** [comment]
Good plan!

**Comment by fgsch** [comment]
This issue has timed out as it has not received any update in over 2 years.
If this is still a problem please open a new issue.
